Userland/Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -148,15 +149,164 @@ void Document::removed_last_ref()
 }
 
 // https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#dom-document-write
-void Document::write(Vector<String> const& strings)
+ExceptionOr<void> Document::write(Vector<String> const& strings)
 {
-    dbgln("TODO: document.write({})", strings);
+    // 1. If document is an XML document, then throw an "InvalidStateError" DOMException.
+    if (doctype() && doctype()->name() == "xml") {
+        return DOM::InvalidStateError::create("Write() called on XML document.");
+    }
+
+    // 2. If document's throw-on-dynamic-markup-insertion counter is greater than 0, then throw an "InvalidStateError" DOMException.
+    if (m_throw_on_dynamic_markup_insertion_counter > 0) {
+        return DOM::InvalidStateError::create("throw-on-dynamic-markup-insertion-counter greater then zero.");
+    }
+
+    // 3. If document's active parser was aborted is true, then return.
+    if (m_active_parser_was_aborted) {
+        return {};
+    }
+
+    // 4. If the insertion point is undefined, then:
+    if (!(m_parser && m_parser->tokenizer()->is_insertion_point_defined())) {
+        //      1. If document's unload counter is greater than 0 or document's ignore-destructive-writes counter is greater than 0, then return.
+        if (m_unload_counter > 0 || m_ignore_destructive_writes_counter > 0) {
+            return {};
+        }
+        //      2. Run the document open st with document.
+        open();
+    }
+
+    // 5. Insert input into the input stream just before the insertion point.
+    StringBuilder builder {};
+    builder.join("", strings);
+    m_parser->tokenizer()->insert_input_at_insertion_point(builder.build());
+
+    // 6. If there is no pending parsing-blocking script, have the HTML parser process input, one code point at a time, processing resulting tokens as they are emitted, and stopping when the tokenizer reaches the insertion point or when the processing of the tokenizer is aborted by the tree construction stage (this can happen if a script end tag token is emitted by the tokenizer).
+    if (!pending_parsing_blocking_script()) {
+        m_parser->run();
+    }
+    return {};
 }
 
 // https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#dom-document-writeln
-void Document::writeln(Vector<String> const& strings)
+ExceptionOr<void> Document::writeln(Vector<String> const& strings)
 {
-    dbgln("TODO: document.writeln({})", strings);
+
+    // FIXME: No need to allocate a new vector
+    Vector<String> new_strings;
+    for (auto const& elem : strings) {
```

Comment:
Please don't use unnecessary abbreviations like `elem`. There is no shortage of horizontal space :)